### PR TITLE
OEP-9: User Authorization (Permissions)

### PR DESCRIPTION
**Pull request by jmbowman**
Presenting an outline of proposed best practices for implementing and using permission checks in Open edX Python code.  Mostly reuses existing APIs and packages, although there is one noted gap for implementing queryset filtering with rule-based permissions.  A rough sketch of a proposed package to fill that gap is provided, although custom model manager methods and Django REST Framework filter classes can be used in the meantime.

### CODE REVIEW

**Comment by maxrothman** [comment]
@bderusha this sounds similar to work you were doing at the hackathon at the last Open edX conference.

**Comment by bderusha** [comment]
@maxrothman is correct. That hackathon project work lives in its entirety in [ECOM-4774](https://openedx.atlassian.net/browse/ECOM-4774) and includes a piece for redesigning how staff permissions work ([ECOM-4841](https://openedx.atlassian.net/browse/ECOM-4841)).  The gist being that staff should not need to be enrolled in a course to access the course content and that the enrollment table is for students.

It sounds like this work would provide a nice framework for where the new rules should live and how the rules around being able to access course content could work.  👍

File: oeps/oep-0009.rst [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+The ``rules`` package can be added to existing packages with minimal impact,
+as it doesn't inherently change the outcome of any authorization checks.  In
+the handful of places where Open edX code actually uses the Django
+authorization API already, the permission implementation can be switched to
+use ``rules`` when convenient (at which point whichever implementation is
+already in use should be deactivated so they don't come to disagree over
+time.)  New code can be written to follow the guidelines in this OEP, and
+existing code can be gradually updated as the need arises.  Effort should
+be made to update all relevant code related to a new explicit permission,
+however, to avoid tempting developers into thinking they've fully updated the
```
Comment (doctoryes) [inline oeps/oep-0009.rst]: Nit: This `however` might be better placed at the beginning of the sentence. It took me a few reads to parse the whole sentence due to the `however` in the middle.

**Comment by doctoryes** [comment]
This standard would be a big improvement for the platform. The code samples greatly help - if adopted, maybe even a sample conversion could be done and documented.
👍

**Comment by efagin** [comment]
@jmbowman I think this OEP is ready to be merged. Do you want to ask for another round of feedback or are you satisfied? (If you are, please update the state to Approved and I'll merge it.)

Thanks for the contribution! This is a really helpful OEP.

**Comment by jmbowman** [comment]
@efagin I think we've gotten about all the feedback we can expect at this point.  Status is updated and it's ready to merge; I removed the copyright footer because the theme provides it as of #28 , so we may want to merge that first.

**Comment by efagin** [comment]
Got it. Thanks - you have my go-ahead to merge after the theme's been
installed.

**Pull request merged**